src/driver/amdxdna/aie2_pci.c

```diff
@@ -583,7 +583,11 @@ static int aie2_init(struct amdxdna_dev *xdna)
 	}
 	xdna->dev_handle = ndev;
 
-	aie2_smu_setup(ndev);
+	ret = aie2_smu_setup(ndev);
+	if (ret) {
+		XDNA_ERR(xdna, "smu setup failed, ret %d", ret);
+		goto disable_sva;
```

Comment:
don't you need to introduce a new goto label? Now you need to power off npu when you fail in below code, right? To make the code symmetric and easy to maintain, you can add aie2_smu_teardown() to undo what is done in aie2_smu_setup().

Reply:
The clean up is being taken care in the aie2_smu_setup.

Reply:
Thanks max for the offline clarification, also I see I can call aie2_smu_set_poweroff to undo what is done in aie2_smu_setup() instead of having a separate function aie2_smu_teardown() which just calls aie2_smu_set_poweroff
.I  hope that's okay

Reply:
Also do we need to set the clks back to min value during the aie2_smu_teardown path?

Reply:
If that is safer to power it off, we should do it. Basically, we should undo all the changes done in aie2_smu_setup(). I think aie2_smu_teardown() is needed to clearly undo all of these in the error path **and** needs to be called in driver's remove code path. Anything asymmetric in the code almost always means something is not right.

Reply:
I have added aie2_smu_teardown() and added the undo functionality of aie2_smu_setup(). 
Onething is I didn't add this aie2_smu_teardown in the driver's remove code path as that's redundant with aie2_hw_stop().
This aie2_hw_stop() is  being  called in aie2_fini() which is setting the dpm level 0 and smu power off.